limiter.go

```diff
@@ -77,17 +77,26 @@ func newDialLimiterWithParams(df dialfunc, fdLimit, perPeerLimit int) *dialLimit
 func (dl *dialLimiter) freeFDToken() {
 	dl.fdConsuming--
 
-	if len(dl.waitingOnFd) > 0 {
+	for len(dl.waitingOnFd) > 0 {
 		next := dl.waitingOnFd[0]
 		dl.waitingOnFd[0] = nil // clear out memory
 		dl.waitingOnFd = dl.waitingOnFd[1:]
+
 		if len(dl.waitingOnFd) == 0 {
-			dl.waitingOnFd = nil // clear out memory
+			// clear out memory.
+			dl.waitingOnFd = nil
+		}
+
+		// Skip over canceled dials instead of queuing up a goroutine.
+		if next.cancelled() {
+			dl.freePeerToken(next)
```

Comment:
Good catch, the peer token wasn't being freed in the case of a cancellation.

Reply:
I don't think *that* was the issue. We'd (a) spawn an `executeDial` goroutine and then (b) free the token when we returned from it. *This* PR just avoids spawning goroutines that we don't need.